File: src/ios/trees/QtSpan.swift
import Foundation

struct QtSpan
{
    var min: QtCoordinate // inclusive
    var max: QtCoordinate // non-inclusive

    static let invalid = QtSpan(min:0, max:0)

    // center point
    var mid: QtCoordinate
    {
        get { return self.min + (self.max - self.min)/2 }
    }

    var length: QtCoordinate
    {
        get { return self.max - self.min }
    }

    func getEnd(_ i: Int) -> QtCoordinate
    {
        assert(i == 0 || i == 1)
        return (i == 0 ? self.min : self.max)
    }

    func getHalf(_ i: Int) -> QtSpan
    {
        assert(i == 0 || i == 1)
        return QtSpan(min: (i == 0 ? self.min : self.mid), max: (i == 0 ? self.mid : self.max)) 
    }

    func inflated(ratio: QtCoordinate) -> QtSpan
    {
        let dl = self.length * (ratio - 1.0)
        return QtSpan(min:self.min-dl, max:self.max+dl)
    }

    func contains(coord: QtCoordinate) -> Bool
    {
        return coord >= self.min && coord < self.max
    }

    func contains(span: QtSpan) -> Bool
    {
        return self.contains(coord:span.min) && self.contains(coord:span.max)
    }

    func intersects(span: QtSpan) -> Bool
    {
        return !(self.max <= span.min || self.min >= span.max)
    }
}

File: src/android/app/src/main/java/com/crickettechnology/trees/QtSpan.java
package com.crickettechnology.trees;

class QtSpan
{
    public QtSpan(double min, double max)
    {
        this.min = min;
        this.max = max;
    }

    public double min; // inclusive
    public double max; // non-inclusive

    public static final QtSpan invalid = new QtSpan(0, 0);

    // center point
    public double getMid()
    {
         return this.min + (this.max - this.min)/2;
    }

    public double getLength()
    {
        return this.max - this.min;
    }

    public QtSpan getHalf(int i)
    {
        assert(i == 0 || i == 1);
        return new QtSpan((i == 0 ? this.min : getMid()), (i == 0 ? getMid() : this.max));
    }

    public QtSpan inflated(double ratio)
    {
        double dl = this.getLength() * (ratio - 1.0);
        return new QtSpan(this.min - dl, this.max + dl);
    }

    public QtSpan getHighHalf()
    {
        return new QtSpan(this.getMid(), this.max);
    }

    public boolean contains(double coord)
    {
        return coord >= this.min && coord < this.max;
    }

    public boolean contains(QtSpan span)
    {
        return this.contains(span.min) && this.contains(span.max);
    }

    public boolean intersects(QtSpan span)
    {
        return !(this.max <= span.min || this.min >= span.max);
    }
}


